File: packages/pigeon/platform_tests/alternate_language_test_plugin/example/ios/RunnerTests/AsyncHandlersTest.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@import Flutter;
@import XCTest;

@import alternate_language_test_plugin;

#import "MockBinaryMessenger.h"

///////////////////////////////////////////////////////////////////////////////////////////
@interface MockHostSmallApi : NSObject <FLTHostSmallApi>
@property(nonatomic, copy) NSString *output;
@property(nonatomic, retain) FlutterError *voidVoidError;
@end

///////////////////////////////////////////////////////////////////////////////////////////
@implementation MockHostSmallApi

- (void)echoString:(NSString *)value
        completion:(nonnull void (^)(NSString *_Nullable, FlutterError *_Nullable))completion {
  if (self.output) {
    completion(self.output, nil);
  } else {
    completion(nil, [FlutterError errorWithCode:@"hey" message:@"ho" details:nil]);
  }
}

- (void)voidVoidWithCompletion:(nonnull void (^)(FlutterError *_Nullable))completion {
  completion(self.voidVoidError);
}

@end

///////////////////////////////////////////////////////////////////////////////////////////
@interface AsyncHandlersTest : XCTestCase
@end

///////////////////////////////////////////////////////////////////////////////////////////
@implementation AsyncHandlersTest

- (void)testAsyncHost2Flutter {
  MockBinaryMessenger *binaryMessenger =
      [[MockBinaryMessenger alloc] initWithCodec:FLTGetCoreTestsCodec()];
  NSString *value = @"Test";
  binaryMessenger.result = value;
  FLTFlutterIntegrationCoreApi *flutterApi =
      [[FLTFlutterIntegrationCoreApi alloc] initWithBinaryMessenger:binaryMessenger];
  XCTestExpectation *expectation = [self expectationWithDescription:@"echo callback"];
  [flutterApi echoAsyncString:value
                   completion:^(NSString *_Nonnull output, FlutterError *_Nullable error) {
                     XCTAssertEqualObjects(output, value);
                     [expectation fulfill];
                   }];
  [self waitForExpectationsWithTimeout:1.0 handler:nil];
}

- (void)testAsyncFlutter2HostVoidVoid {
  MockBinaryMessenger *binaryMessenger =
      [[MockBinaryMessenger alloc] initWithCodec:FLTGetCoreTestsCodec()];
  MockHostSmallApi *mockHostSmallApi = [[MockHostSmallApi alloc] init];
  SetUpFLTHostSmallApi(binaryMessenger, mockHostSmallApi);
  NSString *channelName = @"dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.voidVoid";
  XCTAssertNotNil(binaryMessenger.handlers[channelName]);

  XCTestExpectation *expectation = [self expectationWithDescription:@"voidvoid callback"];
  binaryMessenger.handlers[channelName](nil, ^(NSData *data) {
    NSArray<id> *outputList = [binaryMessenger.codec decode:data];
    XCTAssertEqualObjects(outputList[0], [NSNull null]);
    [expectation fulfill];
  });
  [self waitForExpectationsWithTimeout:1.0 handler:nil];
}

- (void)testAsyncFlutter2HostVoidVoidError {
  MockBinaryMessenger *binaryMessenger =
      [[MockBinaryMessenger alloc] initWithCodec:FLTGetCoreTestsCodec()];
  MockHostSmallApi *mockHostSmallApi = [[MockHostSmallApi alloc] init];
  mockHostSmallApi.voidVoidError = [FlutterError errorWithCode:@"code"
                                                       message:@"message"
                                                       details:nil];
  SetUpFLTHostSmallApi(binaryMessenger, mockHostSmallApi);
  NSString *channelName = @"dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.voidVoid";
  XCTAssertNotNil(binaryMessenger.handlers[channelName]);

  XCTestExpectation *expectation = [self expectationWithDescription:@"voidvoid callback"];
  binaryMessenger.handlers[channelName](nil, ^(NSData *data) {
    NSArray<id> *outputList = [binaryMessenger.codec decode:data];
    XCTAssertNotNil(outputList);
    XCTAssertEqualObjects(outputList[0], mockHostSmallApi.voidVoidError.code);
    [expectation fulfill];
  });
  [self waitForExpectationsWithTimeout:1.0 handler:nil];
}

- (void)testAsyncFlutter2Host {
  MockBinaryMessenger *binaryMessenger =
      [[MockBinaryMessenger alloc] initWithCodec:FLTGetCoreTestsCodec()];
  MockHostSmallApi *mockHostSmallApi = [[MockHostSmallApi alloc] init];
  NSString *value = @"Test";
  mockHostSmallApi.output = value;
  SetUpFLTHostSmallApi(binaryMessenger, mockHostSmallApi);
  NSString *channelName = @"dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.echo";
  XCTAssertNotNil(binaryMessenger.handlers[channelName]);

  NSData *inputEncoded = [binaryMessenger.codec encode:@[ value ]];
  XCTestExpectation *expectation = [self expectationWithDescription:@"echo callback"];
  binaryMessenger.handlers[channelName](inputEncoded, ^(NSData *data) {
    NSArray<id> *outputList = [binaryMessenger.codec decode:data];
    NSString *output = outputList[0];
    XCTAssertEqualObjects(output, value);
    [expectation fulfill];
  });
  [self waitForExpectationsWithTimeout:1.0 handler:nil];
}

- (void)testAsyncFlutter2HostError {
  MockBinaryMessenger *binaryMessenger =
      [[MockBinaryMessenger alloc] initWithCodec:FLTGetCoreTestsCodec()];
  MockHostSmallApi *mockHostSmallApi = [[MockHostSmallApi alloc] init];
  SetUpFLTHostSmallApi(binaryMessenger, mockHostSmallApi);
  NSString *channelName = @"dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.echo";
  XCTAssertNotNil(binaryMessenger.handlers[channelName]);

  NSData *inputEncoded = [binaryMessenger.codec encode:@[ @"Test" ]];
  XCTestExpectation *expectation = [self expectationWithDescription:@"echo callback"];
  binaryMessenger.handlers[channelName](inputEncoded, ^(NSData *data) {
    NSArray<id> *outputList = [binaryMessenger.codec decode:data];
    XCTAssertNotNil(outputList);
    XCTAssertEqualObjects(outputList[0], @"hey");
    XCTAssertEqualObjects(outputList[1], @"ho");
    [expectation fulfill];
  });
  [self waitForExpectationsWithTimeout:1.0 handler:nil];
}

@end

File: packages/pigeon/platform_tests/test_plugin/example/ios/RunnerTests/AsyncHandlersTest.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
import Flutter
import Testing

@testable import test_plugin

class MockHostSmallApi: HostSmallApi {
  var output: String?

  func echo(aString: String) async throws -> String {
    return output!
  }

  func voidVoid() async throws {}
}

@MainActor
struct AsyncHandlersTest {

  @Test
  func asyncHost2Flutter() async throws {
    let value = "Test"
    let binaryMessenger = MockBinaryMessenger<String>(codec: CoreTestsPigeonCodec.shared)
    binaryMessenger.result = value
    let flutterApi = FlutterIntegrationCoreApi(binaryMessenger: binaryMessenger)

    let res = try await flutterApi.echo(value)
    #expect(res == value)
  }

  @Test
  func asyncFlutter2HostVoidVoid() async throws {
    let binaryMessenger = MockBinaryMessenger<String>(
      codec: FlutterStandardMessageCodec.sharedInstance())
    let mockHostSmallApi = MockHostSmallApi()
    HostSmallApiSetup.setUp(binaryMessenger: binaryMessenger, api: mockHostSmallApi)
    let channelName = "dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.voidVoid"
    #expect(binaryMessenger.handlers[channelName] != nil)

    let data = await withCheckedContinuation { continuation in
      binaryMessenger.handlers[channelName]?(nil) { replyData in
        continuation.resume(returning: replyData)
      }
    }
    let outputList = binaryMessenger.codec.decode(data) as? [Any]
    #expect(outputList?.first is NSNull)
  }

  @Test
  func asyncFlutter2Host() async throws {
    let binaryMessenger = MockBinaryMessenger<String>(
      codec: FlutterStandardMessageCodec.sharedInstance())
    let mockHostSmallApi = MockHostSmallApi()
    let value = "Test"
    mockHostSmallApi.output = value
    HostSmallApiSetup.setUp(binaryMessenger: binaryMessenger, api: mockHostSmallApi)
    let channelName = "dev.flutter.pigeon.pigeon_integration_tests.HostSmallApi.echo"
    #expect(binaryMessenger.handlers[channelName] != nil)

    let inputEncoded = binaryMessenger.codec.encode([value])

    let data = await withCheckedContinuation { continuation in
      binaryMessenger.handlers[channelName]?(inputEncoded) { replyData in
        continuation.resume(returning: replyData)
      }
    }
    let outputList = binaryMessenger.codec.decode(data) as? [Any]
    let output = outputList?.first as? String
    #expect(output == value)
  }
}
